tests/click_types.py

```diff
@@ -0,0 +1,163 @@
+        return (str,)
+    if isinstance(param_type, click.types.BoolParamType):
+        return (bool,)
+    elif isinstance(param_type, (click.types.IntParamType, click.IntRange)):
+        return (int,)
+    elif isinstance(param_type, (click.types.FloatParamType, click.FloatRange)):
+        return (float,)
+    elif isinstance(param_type, click.Choice):
+        return (t.Literal[tuple(param_type.choices)],)
+    elif isinstance(param_type, click.DateTime):
+        return (datetime.datetime,)
+    if isinstance(param_type, click.types.UUIDParameterType):
+        return (uuid.UUID,)
+    if isinstance(param_type, click.File):
+        return (t.TextIO,)
+    if isinstance(param_type, click.Path):
+        if param_type.path_type is not None:
+            if isinstance(param_obj.path_type, type):
+                return (param_obj.path_type,)
+            else:
+                raise NotImplementedError(
+                    "todo: support the return type of a converter func"
+                )
+        else:
+            return (str,)
+
+    # globus-cli types
+    if isinstance(param_type, CommaDelimitedList):
+        return (list[str],)
+    elif isinstance(param_type, EndpointPlusPath):
+        if param_type.path_required:
+            return (tuple[uuid.UUID, str],)
+        else:
+            return (tuple[uuid.UUID, str | None],)
+    elif isinstance(param_type, IdentityType):
+        return (ParsedIdentity,)
+    elif isinstance(param_type, JSONStringOrFile):
+        return (JsonValue,)
+    elif isinstance(param_type, LocationType):
+        return (tuple[float, float],)
+    elif isinstance(param_type, StringOrNull):
+        return (str, ExplicitNullType)
+    elif isinstance(param_type, UrlOrNull):
+        return (str, ExplicitNullType)
+    elif isinstance(param_type, TaskPath):
+        return (TaskPath,)
+    elif isinstance(param_type, NotificationParamType):
+        return (dict[str, bool],)
+    elif isinstance(param_type, TimedeltaType):
+        if param_type._convert_to_seconds:
+            return (int,)
+        return (datetime.timedelta,)
+
+    raise NotImplementedError(f"unsupported parameter type: {param_type}")
+
+
+_NEVER_NULL_CALLBACKS = (none_to_empty_dict,)
+
+
+def _option_defaults_to_none(o: click.Option) -> bool:
+    if o.default is not None:
+        return False
+    if o.callback is None:
+        return True
+    if o.callback in _NEVER_NULL_CALLBACKS:
+        return False
+    return True
+
+
+def check_has_correct_annotations_for_click_args(f):
+    hints = t.get_type_hints(f.callback)
+    param_map = {}
+    for param in f.params:
+        # skip params which do not get passed to the callback
+        if param.expose_value is False:
+            continue
+
+        param_map[param.name] = param
+
+    for name in param_map:
+        if name not in hints:
+            raise ValueError(f"expected parameter '{name}' was not in type hints")
```

Comment:
Yep, this seems like a reasonable step!

I also might _remove_ `param_map`. I'm not sure if it's necessary to the logic, and it might be clearer to look repeatedly at `f.params`.
So it's possible that I'll do more extensive cleanup in a little bit.

Reply:
The refactor I did was more extensive, but should cover this.